bpzpipe/BPZPipe2.py

```diff
@@ -0,0 +1,351 @@
+
+            # Clip to min mag errors
+            np.clip(mag_errs, self.config['mag_err_min'], 1e10, mag_errs)
+
+            # Convert to pseudo-fluxes
+            flux = 10.0**(-0.4*mags)
+            flux_err = flux * (10.0**(0.4*mag_errs) - 1.0)
+
+            # Check if an object is seen in each band at all.
+            # Fluxes not seen at all are listed as infinity in the input,
+            # so will come out as zero flux and zero flux_err.
+            # Check which is which here, to use with the ZP errors below
+            seen1 = (flux > 0) & (flux_err > 0)
+            seen = np.where(seen1)
+            unseen = np.where(~seen1)
+
+            # Add zero point magnitude errors.
+            # In the case that the object is detected, this
+            # correction depends onthe flux.  If it is not detected
+            # then BPZ uses half the errors instead
+            add_err = np.zeros_like(flux_err)
+            add_err[seen] = ((zp_frac*flux)**2)[seen]
+            add_err[unseen] = ((zp_frac*0.5*flux_err)**2)[unseen]
+            flux_err = np.sqrt(flux_err**2 + add_err)
+
+            # Upate the input dictionary with new things we have calculated
+            data[f'flux{suffix}'] = flux
+            data[f'flux_err{suffix}'] = flux_err
+            data[f'mags{suffix}'] = mags
+
+    def estimate_pdfs(self, flux_templates, z, data):
+
+        # BPZ uses the magnitude in one band to get a prior.
+        # Select which one here.
+        bands = self.config['bands']
+        prior_band = self.config['prior_band']
+        m_0_col = bands.index(prior_band)
+
+        # Size of output
+        ng = len(data['mags'])
+        nz = len(z)
+
+        # Optionally, BPZ can convolve the PDF with a Gaussian.
+        # This consolidates multiple nearby peaks into one, which is
+        # useful for estimating the MAP point, among other things.
+        sigma_g = self.config['sigma_g']
+
+        # Prepare said Gaussian
+        if sigma_g > 0:
+            dz = self.config['dz']
+            x = np.arange(-3.*sigma_g, 3.*sigma_g + dz/10., dz)
+            kernel = np.exp(-(x/sigma_g)**2)
+        else:
+            kernel = None
+
+
+        # Space for the output
+        pdfs = np.zeros((ng, nz))
+        point_estimates = np.zeros((5, ng))
+
+        # Metacal variants
+        suffices = ["", "_1p", "_1m", "_2p", "_2m"]
+        for s, suffix in enumerate(suffices):
+            for i in range(ng):
+                # Pull out the rows of data for this galaxy
+                mag_0 = data[f'mags{suffix}'][i, m_0_col]
+                flux  = data[f'flux{suffix}'][i]
+                flux_err = data[f'flux_err{suffix}'][i]
+
+                # and compute the PDF for it
+                pdf = self.estimate_pdf(flux_templates, kernel, flux, flux_err, mag_0, z)
+                
+                if suffix=="":
+                    pdfs[i] = pdf
+
+                # The pdf already sums to unity, so this gives us the mean
```

Comment:
The default point estimate for BPZ is the maximum value of the marginalized posterior, not the mean.  If you want to use the mean in place of the mode that's fine, but for testing and comparison to previous BPZ runs I replaced the following line of code with 
`
point_estimates[s,i] = z[np.argmax(pdf)]
`

Reply:
I'll make this an option in the input.

Reply:
Done & pushed.